crates/ruff/src/commands/format.rs

```diff
@@ -628,6 +739,127 @@ impl<'a> FormatResults<'a> {
+            };
+
+            let mut diagnostic = Diagnostic::new(
+                DiagnosticId::Unformatted,
+                Severity::Error,
+                "File would be reformatted",
+            );
+
+            let path = result.path.to_string_lossy();
+            // For now, report the edit as a replacement of the whole file's contents. For
+            // scripts, this is a single `Edit`, but notebook edits must be split by cell in
+            // order to render them as diffs.
+            //
+            // We also attempt to estimate the line number width for aligning the
+            // annotate-snippets header. This is only an estimate because we don't actually know
+            // if the maximum line number present in the document will be rendered as part of
+            // the diff, either as a changed line or as an unchanged context line. For
+            // notebooks, we refine our estimate by checking the number of lines in each cell
+            // individually, otherwise we could use `formatted.source_code().count_lines(...)`
+            // in both cases.
+            let (fix, line_count) = if let SourceKind::IpyNotebook(formatted) = formatted
+                && let SourceKind::IpyNotebook(unformatted) = unformatted
+            {
+                notebook_index.insert(path.to_string(), unformatted.index().clone());
+
+                let mut edits = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .zip(unformatted.cell_offsets().ranges())
+                    .map(|(formatted_range, unformatted_range)| {
+                        let formatted = &formatted.source_code()[formatted_range];
+                        Edit::range_replacement(formatted.to_string(), unformatted_range)
+                    });
+
+                let fix = Fix::safe_edits(
+                    edits
+                        .next()
+                        .expect("Formatted files must have at least one edit"),
+                    edits,
+                );
+                let source = formatted.source_code();
+                let line_count = formatted
+                    .cell_offsets()
+                    .ranges()
+                    .map(|range| source.count_lines(range))
+                    .max()
+                    .unwrap_or_default();
+                (fix, line_count)
+            } else {
+                let fix = Fix::safe_edit(Edit::range_replacement(
+                    formatted.source_code().to_string(),
+                    TextRange::up_to(unformatted.source_code().text_len()),
+                ));
+                let line_count = formatted
+                    .source_code()
+                    .count_lines(TextRange::up_to(formatted.source_code().text_len()));
+                (fix, line_count)
+            };
+
+            // Locate the first and last characters that differ to use as the diagnostic
+            // range.
+            let range = {
```

Comment:
I think we can do better than traversing the entire file by:

1. Search from the beginning and find the first changed character
2. Slice the remaining document to `&source[start..]`
3. Search from the end to find the last changed character


This is still `O(n)` in the worst case, but should do much better if a large portion of the file changed. It might be worth extracting this logic into a helper function to not distract the reading flow